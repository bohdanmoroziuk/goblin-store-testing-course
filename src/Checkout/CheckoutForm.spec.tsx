import React from "react"
import { render, fireEvent } from "@testing-library/react";
import { act } from "react-dom/test-utils";

import { CheckoutForm } from "./CheckoutForm";

describe("CheckoutForm", () => {
  it("renders correctly", () => {
    const { container } = render(<CheckoutForm />);

    expect(container.innerHTML).toMatch("Cardholder's Name");
    expect(container.innerHTML).toMatch("Card Number");
    expect(container.innerHTML).toMatch("Expiration Date");
    expect(container.innerHTML).toMatch("CVV");
    expect(container.innerHTML).toMatch("Place order");
  })

  describe("with validation errors", () => {
    it("renders error messages", async () => {
      const { container, getByText } = render(<CheckoutForm />);

      await act(async () => {
        fireEvent.click(getByText("Place order"));
      });

      expect(container.innerHTML).toMatch("name is a required field");
      expect(container.innerHTML).toMatch("cardNumber is a required field");
      expect(container.innerHTML).toMatch("expDate must be a `date` type, but the final value was: `Invalid Date` (cast from the value `\"\"`).");
      expect(container.innerHTML).toMatch("cvv is a required field");
    })
    it.todo("disables submit button")
  })

  describe("without validation errors", () => {
    describe("on place order button click", () => {
      it.skip("calls submit function with form data", async () => {
        const mockSubmit = jest.fn();
        
        const { getByText, getByPlaceholderText, getByLabelText } = render(<CheckoutForm submit={mockSubmit} />);

        await act(async () => {
          fireEvent.change(
            getByPlaceholderText("John Smith"),
            { target: { value: "Bibo Bobbins" } },
          );
          fireEvent.change(
            getByPlaceholderText("0000 0000 0000 0000"),
            { target: { value: "0000 0000 0000 0000" } },
          );
          fireEvent.change(
            getByLabelText("Expiration Date"),
            { target: { value: "3020-05" } },
          );
          fireEvent.change(
            getByPlaceholderText("000"),
            { target: { value: "123" } },
          );
        });

        await act(async () => {
          fireEvent.click(getByText("Place order"));
        });

        expect(mockSubmit).toHaveBeenCalled();
      })
      it.todo("clears cart")
      it.todo("redirects to order summary page")
    })
  })
})
