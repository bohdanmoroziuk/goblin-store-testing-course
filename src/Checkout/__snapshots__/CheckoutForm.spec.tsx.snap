// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutForm renders correctly 1`] = `
<div>
  <form>
    <div
      class="nes-field"
    >
      <label
        for="name"
      >
        Cardholder's Name
        :
      </label>
      <input
        class="nes-input undefined"
        name="name"
        placeholder="John Smith"
        type="text"
      />
    </div>
    <div
      class="nes-field"
    >
      <label
        for="cardNumber"
      >
        Card Number
        :
      </label>
      <input
        autocomplete="cc-number"
        class="nes-input undefined"
        inputmode="numeric"
        name="cardNumber"
        placeholder="0000 0000 0000 0000"
        type="tel"
      />
    </div>
    <div
      class="nes-field"
    >
      <label
        for="expDate"
      >
        Expiration Date
        :
      </label>
      <input
        class="nes-input undefined"
        name="expDate"
        type="month"
      />
    </div>
    <div
      class="nes-field"
    >
      <label
        for="cvv"
      >
        CVV
        :
      </label>
      <input
        class="nes-input undefined"
        name="cvv"
        placeholder="000"
        type="number"
      />
    </div>
    <button
      class="nes-btn is-primary"
    >
      Place order
    </button>
  </form>
</div>
`;
